packages/expo/cli/start/server/UrlCreator.ts

```diff
@@ -0,0 +1,159 @@
+import assert from 'assert';
+import { URL } from 'url';
+
+import * as Log from '../../log';
+import { getIpAddress } from '../../utils/ip';
+
+export interface CreateURLOptions {
+  /** URL scheme to use when opening apps in custom runtimes. */
+  scheme?: string | null;
+  /** Type of dev server host to use. */
+  hostType?: 'localhost' | 'lan' | 'tunnel';
+  /** Requested hostname. */
+  hostname?: string;
+}
+
+export class UrlCreator {
+  constructor(
+    private defaults: CreateURLOptions,
+    private bundlerInfo: { port: number; getTunnelUrl?: () => string }
+  ) {}
+
+  /**
+   * @returns URL like `http://localhost:19000/_expo/loading?platform=ios`
+   */
+  public constructLoadingUrl(opts: CreateURLOptions, platform: string) {
+    const url = new URL('_expo/loading', this.constructUrl({ scheme: 'http', ...opts }));
+    url.search = new URLSearchParams({ platform }).toString();
+    return url.toString();
+  }
+
+  /** Create a URI for launching in a native dev client. Returns `null` when no `scheme` can be resolved. */
+  public constructDevClientUrl(opts?: CreateURLOptions): null | string {
+    const protocol: string = opts?.scheme || this.defaults.scheme;
+
+    if (
+      !protocol ||
+      // Prohibit the use of http(s) in dev client URIs since they'll never be valid.
```

Comment:
Warn in this case?

Reply:
In this case we will fallback on bundle identifier, launch manifest, etc.